### PR TITLE
Test/jump with time range

### PR DESCRIPTION
**Pull request by mudhoney**
Implement "Jump Time Range" tests, 
Implement convenience functions for screenshots

### CODE REVIEW

File: tests/desktop/normal/observation_date/jump_with_time_range.spec.ts [hunk 1]
```diff
@@ -0,0 +1,167 @@
+import { test, expect } from '@playwright/test';
+import { Helioviewer } from '../../../page_objects/helioviewer';
+import path from 'path';
```
Comment (dgarciabriseno) [inline tests/desktop/normal/observation_date/jump_with_time_range.spec.ts]: Unused import
```suggestion
```

File: tests/page_objects/helioviewer.ts [hunk 2]
```diff
@@ -312,6 +328,89 @@ class Helioviewer {
         await this.page.locator('#logo').hover();
     }
 
+    /**
+    * Get the loaded date in helioviewer 
+    * @returns Date|null - Loaded date of helioviewer, it can be null if any error.
+    */
+    async GetLoadedDate(): Date|null {
```
Comment (dgarciabriseno) [inline tests/page_objects/helioviewer.ts]: async functions always return type `Promise`
```suggestion
    async GetLoadedDate(): Promise<Date|null> {
```

File: tests/page_objects/helioviewer.ts [hunk 3]
```diff
@@ -312,6 +328,89 @@ class Helioviewer {
...
+
+        const currentDate = await this.page.getByLabel('Observation date', {exact:true}).inputValue();
+        const currentTime = await this.page.getByLabel('Observation time', {exact:true}).inputValue();
+
+        const date = new Date(currentDate +' '+ currentTime);
+
+        if (isNaN(date.getTime())) {
+            return null;
+        }
```
Comment (dgarciabriseno) [inline tests/page_objects/helioviewer.ts]: This is bad, right? Should we just move the expect here?

```suggestion
await expect(date.getTime()).not.toBeNull();
```

File: tests/desktop/normal/observation_date/jump_with_time_range.spec.ts [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+      await hv.CloseAllNotifications();
+
+      // 3. USE NEWEST SOHO
+      await hv.UseNewestImage();
+      await hv.WaitForLoadingComplete();
+      await hv.CloseAllNotifications();
+
+      // 4. MARK TIME BEFORE JUMP BACKWARDS
+      const dateBeforeJump = await hv.GetLoadedDate();
+      expect(dateBeforeJump).not.toBeNull();
```
Comment (dgarciabriseno) [inline tests/desktop/normal/observation_date/jump_with_time_range.spec.ts]: You can remove this if you make GetLoadedDate assert internally.

File: tests/desktop/normal/observation_date/jump_with_time_range.spec.ts [hunk 5]
```diff
@@ -0,0 +1,167 @@
+import { test, expect } from '@playwright/test';
+import { Helioviewer } from '../../../page_objects/helioviewer';
+import path from 'path';
+
+const time_jump_ranges = [
+  { jump_label: '1Min', seconds : 60 }, 
+  { jump_label: '6Hours', seconds : 21600 }, 
+  { jump_label: '1Week', seconds : 604800 }, 
+  { jump_label: '1Year', seconds : 31556926 }, 
+];
```
Comment (dgarciabriseno) [inline tests/desktop/normal/observation_date/jump_with_time_range.spec.ts:9]: Should we just test all options?

File: tests/page_objects/helioviewer.ts [hunk 6]
```diff
@@ -312,6 +328,89 @@ class Helioviewer {
...
+    */
+    async GetLoadedDate(): Date|null {
+
+        const currentDate = await this.page.getByLabel('Observation date', {exact:true}).inputValue();
+        const currentTime = await this.page.getByLabel('Observation time', {exact:true}).inputValue();
+
+        const date = new Date(currentDate +' '+ currentTime);
+
+        if (isNaN(date.getTime())) {
+            return null;
```
Comment (mudhoney) [inline tests/page_objects/helioviewer.ts]: fail here test.not.null

File: tests/desktop/normal/observation_date/jump_with_time_range.spec.ts [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+      });
+
+      // 14, 2 SCREENSHOTS ARE FROM SAME DATE, AND SHOULD MATCH
+      await expect(directDateScreenshot).toBe(afterJumpScreenshot)
+    });
+
+    /**
+     * This test is testing jumping forward functionality with given label for range select-box 
+     */
+    test('Jump forwards with '+jump_label+' should go to matching datetime in future, with matching screenshots', async ({ page, context, browser }, info) => {
```
Comment (mudhoney) [inline tests/desktop/normal/observation_date/jump_with_time_range.spec.ts]: production tags

File: tests/page_objects/helioviewer.ts [hunk 8]
```diff
@@ -2,13 +2,14 @@
  * @file Contains functions for interacting with the Helioviewer UI
  */
 
-import { Locator, Page, expect } from '@playwright/test';
+import { Locator, Page, Info, expect } from '@playwright/test';
```
Comment (dgarciabriseno) [inline tests/page_objects/helioviewer.ts]:
```suggestion
import { Locator, Page, TestInfo, expect } from '@playwright/test';
```

File: tests/page_objects/helioviewer.ts [hunk 9]
```diff
@@ -2,13 +2,14 @@
...
 import { ImageLayer } from './image_layer';
 import { Screenshot } from './screenshot';
 import { Movie } from './movie';
 import { URLShare } from './urlshare';
 import { EventTree } from './event_tree';
 import { VSODrawer } from './vso_drawer';
+import * as fs from 'fs';
```
Comment (dgarciabriseno) [inline tests/page_objects/helioviewer.ts:12]:
```suggestion
```

File: tests/page_objects/helioviewer.ts [hunk 10]
```diff
@@ -312,6 +328,87 @@ class Helioviewer {
         await this.page.locator('#logo').hover();
     }
 
+    /**
+    * Get the loaded date in helioviewer 
+    * @returns Date|null - Loaded date of helioviewer, it can be null if any error.
+    */
+    async GetLoadedDate(): Promise<Date|null> {
```
Comment (dgarciabriseno) [inline tests/page_objects/helioviewer.ts]:
```suggestion
    async GetLoadedDate(): Promise<Date> {
```

File: tests/page_objects/helioviewer.ts [hunk 11]
```diff
@@ -312,6 +328,87 @@ class Helioviewer {
...
+
+        return date
+    }
+
+    /**
+    * Jump backwards with jump button, with given seconds layer 
+    * @param integer seconds, interval in seconds 
+    * @returns void
+    */
+    async JumpBackwardsDateWithSelection(seconds: integer): void {
```
Comment (dgarciabriseno) [inline tests/page_objects/helioviewer.ts]:
```suggestion
    async JumpBackwardsDateWithSelection(seconds: number): Promise<void> {
```

File: tests/page_objects/helioviewer.ts [hunk 12]
```diff
@@ -312,6 +328,87 @@ class Helioviewer {
...
+        await this.page.getByLabel('Jump:').selectOption(seconds.toString());
+        await this.page.locator('#timeBackBtn').click();
+    }
+
+    /**
+    * Jump forward with jump button, with given seconds layer 
+    * @param integer seconds, interval in seconds 
+    * @returns void
+    */
+    async JumpForwardDateWithSelection(seconds: integer): void {
```
Comment (dgarciabriseno) [inline tests/page_objects/helioviewer.ts]:
```suggestion
    async JumpForwardDateWithSelection(seconds: number): Promise<void> {
```

File: tests/page_objects/helioviewer.ts [hunk 13]
```diff
@@ -286,22 +288,36 @@ class Helioviewer {
         await this.WaitForLoadingComplete();
     }
 
+    /**
+    * Sets observation datetime of Helioviewer from given Date object, 
+    * @param Date date - The date object , to be used to load observation datetime.
+    * @returns void 
+    */
+    async SetObservationDateTimeFromDate(date: Date): void {
```
Comment (dgarciabriseno) [inline tests/page_objects/helioviewer.ts]:
```suggestion
    async SetObservationDateTimeFromDate(date: Date): Promise<void> {
```

File: tests/page_objects/helioviewer.ts [hunk 14]
```diff
@@ -312,6 +328,87 @@ class Helioviewer {
...
+        return base64Image;
+    }
+
+    /**
+    * Attach base64 screnshot with a given filename to trace report 
+    * and  exit afterwards with false assertion
+    * @param filename string | name of file in trace report
+    * @param options Object | pass options to playwright screenshot function  
+    * @returns void  */
+    async saveScreenshotAndExit(filename: string = "", options: Object = {}): void {
```
Comment (dgarciabriseno) [inline tests/page_objects/helioviewer.ts]:
```suggestion
    async saveScreenshotAndExit(filename: string = "", options: Object = {}): Promise<void> {
```

File: tests/page_objects/helioviewer.ts [hunk 15]
```diff
@@ -312,6 +328,87 @@ class Helioviewer {
...
+    /**
+    * Get the loaded date in helioviewer 
+    * @returns Date|null - Loaded date of helioviewer, it can be null if any error.
+    */
+    async GetLoadedDate(): Promise<Date> {
+
+        const currentDate = await this.page.getByLabel('Observation date', {exact:true}).inputValue();
+        const currentTime = await this.page.getByLabel('Observation time', {exact:true}).inputValue();
+
+        const date = new Date(currentDate +' '+ currentTime);
```
Comment (dgarciabriseno) [inline tests/page_objects/helioviewer.ts]:
```suggestion
        const date = new Date(currentDate +' '+ currentTime + 'Z');
```

**Review by dgarciabriseno — APPROVED** [review]

**Pull request merged**